node/pkg/driver/device_connectivity/device_connectivity_helper_scsigeneric.go

```diff
@@ -505,48 +587,59 @@ func (o OsDeviceConnectivityHelperGeneric) ReloadMultipath() error {
 	return nil
 }
 
-func (o OsDeviceConnectivityHelperGeneric) GetDmsPath(volumeId string, volumeNguid string) (string, error) {
-	volumeUuidLower := strings.ToLower(volumeId)
-
-	mpathdOutput, err := o.Helper.WaitForDmToExist(volumeUuidLower, volumeNguid, WaitForMpathRetries, WaitForMpathWaitIntervalSec)
-
+func (o OsDeviceConnectivityHelperGeneric) GetMpathDeviceName(volumePath string) (string, error) {
+	procMountsFileContent, err := ioutil.ReadFile(procMountsFilePath)
 	if err != nil {
 		return "", err
 	}
+	procMounts := string(procMountsFileContent)
 
-	if mpathdOutput == "" {
-		return "", &MultipathDeviceNotFoundForVolumeError{volumeId}
-	}
+	return o.Helper.GetMpathDeviceNameFromProcMounts(procMounts, volumePath)
+}
 
-	dms := make(map[string]bool)
-	scanner := bufio.NewScanner(strings.NewReader(mpathdOutput))
-	for scanner.Scan() {
-		deviceLine := scanner.Text()
-		lineParts := strings.Split(deviceLine, mpathdSeparator)
-		dm, uuid := lineParts[0], lineParts[1]
-		if strings.Contains(uuid, volumeUuidLower) || strings.Contains(uuid, volumeNguid) {
-			dmPath := filepath.Join(DevPath, filepath.Base(strings.TrimSpace(dm)))
-			dms[dmPath] = true
-			logger.Infof("GetMpathDevice: DM found: %s for volume %s", dmPath, uuid)
-		}
+func (o OsDeviceConnectivityHelperGeneric) GetMpathOutputByVolumeId(volumeUuidLower string, volumeNguid string) (string, error) {
+	mpathdOutput, err := o.Helper.GetMpathOutput(volumeUuidLower, volumeNguid, multipathdFilterMpathNameAndVolumeId)
+	if err != nil {
+		return "", err
 	}
+	return mpathdOutput, nil
+}
 
-	if len(dms) > 1 {
-		return "", &MultipleDmDevicesError{volumeId, dms}
+func (o OsDeviceConnectivityHelperGeneric) GetVolumeIdByMpathName(mpathDeviceName string, mpathdOutput string) (string, error) {
+	identifiers := []string{mpathDeviceName}
+	volumeIds := o.Helper.ParseFieldValuesOfIdentifiers(identifiers, mpathdOutput)
+	if len(volumeIds) > 1 {
+		return "", &MultipleVolumeIdsError{volumeIds, mpathDeviceName}
 	}
 
-	var dm string
-	for dm = range dms {
+	var volumeId string
+	for volumeId = range volumeIds {
 		break // because its a single value in the map(1 mpath device, if not it should fail above), so just take the first
 	}
 
-	return dm, nil
+	return volumeId, nil
+}
+
+func (o OsDeviceConnectivityHelperGeneric) GetDmsPath(volumeId string, volumeNguid string, multipathdCommandFormatArgs []string) (string, error) {
```

Comment:
where is `GetDmsPath` called without `MultipathdFilterMpathAndVolumeId`?

Reply:
yes you are right, I will remove `multipathdCommandFormatArgs` parameter from `GetDmsPath` function and pass there `MultipathdFilterMpathAndVolumeId` to `GetMpathdOutput` function

Reply:
done